005-OFEP-provider-hook.md

```diff
@@ -0,0 +1,49 @@
+## 005-OFEP-provider-hook.md
+
+## State: DRAFTING
+
+As a part of the Provider implementation by the Flag Management Systen, there is a need to have standardized support for hooks specific to the Provider. The 
+Provider Hook(s) should be specific to the provider implementation and implement the various defined stages of the Hook interface as applicable to the 
+specific provider. 
+
+## Assumptions
+
+The Provider Hook is transparent to the application developer. The Provider may choose to document certain default behavior/feature in specific stages of the 
+Provider Hook as relevant to the application and the application developer may be allowed to override the default feature. 
+
+## Example Use Case
+
+Flag Management System X provides the option to add Experimentation as an extension of Feature Flag. The user creates a Feature Flag, runs basic on/off 
+evaluations of the Feature targeting specific audience and then wants to run A/B tests and hence adds the Experimentation aspects to the Feature Flag.
+
+The Provider needs to generate event specific to the Experimentation implementation  on successful flag evaluation. The system wants to use a Provider Hook 
+to do the following for each evaluation  
+1. Run initialization operations (Use the _before_ stage of the Hook)
+2. Run an event publish operation on successful flag evaluation operation (Use the _after_ stage of the Hook)
+3. Provide the option to the application developer to override the event publish operation 
+   1. The application developer may choose not to immediately publish the event on evaluation, but decide to publish it (or not) later based on other 
+      application specific logic
+
+For applications not required to override the default behavior, there is no extra step needed to add any other hooks. The Provider Hook is transparent to 
+the application developer.
+
+## Example Workflow
+
+1. At the time of registering the provider (OpenFeature.setProvider), Provider implementation can register one or more Provider Hook(s)
+2. Provider Hook implements the _before_ and _after_ and any other relevant stages to run the relevant provider logic.
+3. Provider documents the relevant default behavior (e.g. automatic publish of Experimentation event on successful evaluation)
+4. Provider provides an implementation of an Invocation Hook to override this default behavior as a part of the Provider SDK. 
+5. If needed, the application developer uses the invocation hook to override specific default behavior of the provider by using Hook Hints e.g. turn off 
+   default event publish behavior
+
+## Hook Ordering
+
+The Hooks should be executed in specific order.
+
+* before: API, Client, Invocation, Provider
```

Comment:
I dislike the lack of symmetry here. Would it be reasonable to have providers run their before hooks first? I'd love to maintain a simple "the after is the reverse order of before" if we can.

Reply:
In the 'before' stage, IMO, the Provider should run last. This is because in this stage, the context is mutable and the Provider should be responsible for examining the final form of the context and then run the evaluation.
I am not clear on why the hooks should run the reverse order in the 'after' stage. I am curious about the use cases that rationalizes this.

Reply:
> In the 'before' stage, IMO, the Provider should run last. This is because in this stage, the context is mutable and the Provider should be responsible for examining the final form of the context and then run the evaluation.

Agreed.

> I am not clear on why the hooks should run the reverse order in the 'after' stage. I am curious about the use cases that rationalizes this.

Especially for global hooks, having them run in stack order makes a lot of sense - for example, with the OpenTelementry hook, we want to start the span as early as possible and close it at the very end, after all hooks have run, for maximum visibility.

I think in all cases, because these are provider hooks, they should run as close to the provider as possible, so:

```
before: API, Client, Invocation, Provider
after/error/finally: Provider, Invocation, Client, API
```

Reply:
Food for thought, if the ordering of hooks is
```
before: API, Client, Invocation, Provider
after/error/finally: Provider, Invocation, Client, API
```
the only hook that would allow the provider customisability beyond what is already available is the `finally` hook.
A provider already has direct control over what happens immediately before and after a flag call. Whereas there is currently no way for a provider to hook (`finally`) into the execution post success/failure.

Reply:
>A provider already has direct control over what happens immediately before and after a flag call. Whereas there is currently no way for a provider to hook (finally) into the execution post success/failure.

I think this is good, TBH, that's the business of the application author.

What there currently is _not_, is the ability to easily run arbitrary code before/after flag evaluation in a provider, which can lead to a lot of duplicated code in a provider for each flag resolution type (the context-transformer was a way to avoid this, but it has not been widely adopted and only handled one specific use case).

Reply:
I have modified the after hook order as discussed above. 
I like the idea of running the Provider hook  after the invocation in the 'finally' phase to but this will require more logic in the SDK to ensure the following order for finally: Invocation, Provider, Client, API.